src/layer.rs

```diff
@@ -9,58 +13,112 @@ use tracing_subscriber::Layer;
 /// OpenTelemetry layer for use in a project that uses [tracing].
 ///
 /// [tracing]: https://github.com/tokio-rs/tracing
-pub struct OpentelemetryLayer<T: api::Tracer> {
+pub struct OpenTelemetryLayer<S, T: api::Tracer> {
     tracer: T,
+
+    get_context: WithContext,
+    _registry: marker::PhantomData<S>,
 }
 
-struct SpanEventVisitor<'a, S: api::Span>(&'a mut S);
+// this function "remembers" the types of the subscriber so that we
+// can downcast to something aware of them without knowing those
+// types at the callsite.
+//
+// See https://github.com/tokio-rs/tracing/blob/4dad420ee1d4607bad79270c1520673fa6266a3d/tracing-error/src/layer.rs
+pub(crate) struct WithContext(
+    fn(&tracing::Dispatch, &span::Id, f: &mut dyn FnMut(&mut OpenTelemetrySpanInfo)),
+);
+
+impl WithContext {
+    // This function allows a function to be called in the context of the
+    // "remembered" subscriber.
+    pub(crate) fn with_context<'a>(
+        &self,
+        dispatch: &'a tracing::Dispatch,
+        id: &span::Id,
+        mut f: impl FnMut(&mut OpenTelemetrySpanInfo),
+    ) {
+        (self.0)(dispatch, id, &mut f)
+    }
+}
 
-impl<'a, S: api::Span> field::Visit for SpanEventVisitor<'a, S> {
+/// Span extension for tracking OpenTelemetry data
+pub(crate) struct OpenTelemetrySpanInfo {
+    /// Managing a trace id here is necessary for linking spans that do not
+    /// have an external parent.
+    pub(crate) trace_id: api::TraceId,
+    /// SpanBuilder can be used to accumulate state during tracing operations
+    pub(crate) builder: api::SpanBuilder,
+}
+
+pub(crate) fn build_context(otel_info: &mut OpenTelemetrySpanInfo) -> api::SpanContext {
+    let span_id = otel_info.builder.span_id.expect("Builders must have id");
+    let (trace_id, trace_flags) = otel_info
+        .builder
+        .parent_context
+        .as_ref()
+        .map(|parent_context| (parent_context.trace_id(), parent_context.trace_flags()))
+        .unwrap_or((otel_info.trace_id, api::TRACE_FLAG_SAMPLED));
```

Comment:
Should this query the Sampler instead of just defaulting to 100% sampling?

Reply:
Hm great point, the layer could have a clone of the sampler and make decisions here as well.

Reply:
Hmm, it seems like it would be better if you could reuse the Tracer::build() code, since it handles all of the different outcomes of different sampling decisions. Why not create the Span here instead of in on_close?

Reply:
That is an option, there are cases where users of this api may find it confusing if they cannot set the parent after reading the context though. (The builder usage here is necessary because many aspects of the span cannot be modified after creation)

Reply:
Another potential issue is that getting access to the tracer or a potentially added sampler is difficult from this place. I'll open an issue for this.